File: apps/backend/utils/chat_constants.js
/**
 * utils/chat_constants.js
 * -----------------------
 * WHAT:
 * - Shared chat constants for message types, attachment limits, and socket events.
 *
 * WHY:
 * - Keeps chat rules consistent across services, controllers, and sockets.
 * - Prevents magic strings/numbers drifting between frontend and backend.
 *
 * HOW:
 * - Exposes enums + limit values for reuse in chat models + services.
 * - Centralizes event names for Socket.IO emit/subscribe.
 */

// WHY: Use a single source for chat conversation types.
const CHAT_CONVERSATION_TYPES = {
  DIRECT: 'direct',
  GROUP: 'group',
};

// WHY: Standardize message types so parsing stays predictable.
const CHAT_MESSAGE_TYPES = {
  TEXT: 'text',
  ATTACHMENT: 'attachment',
  SYSTEM: 'system',
};

// WHY: Keep call media modes ready for audio now and video later.
const CHAT_CALL_MEDIA_MODES = {
  AUDIO: 'audio',
  VIDEO: 'video',
};

// WHY: Shared call session states keep backend + frontend aligned.
const CHAT_CALL_STATES = {
  RINGING: 'ringing',
  ACTIVE: 'active',
  DECLINED: 'declined',
  ENDED: 'ended',
  MISSED: 'missed',
  CANCELLED: 'cancelled',
};

// WHY: Keep attachment type names consistent for validation.
const CHAT_ATTACHMENT_TYPES = {
  IMAGE: 'image',
  DOCUMENT: 'document',
  AUDIO: 'audio',
};

// WHY: Centralize size limits to avoid inline magic numbers.
const CHAT_LIMITS = {
  MAX_IMAGE_BYTES: 10 * 1024 * 1024,
  MAX_DOCUMENT_BYTES: 20 * 1024 * 1024,
  MAX_AUDIO_BYTES: 16 * 1024 * 1024,
  MAX_TEXT_LENGTH: 4000,
  MAX_TITLE_LENGTH: 120,
};

// WHY: Enumerate supported mime types for validation.
const CHAT_ATTACHMENT_MIME_TYPES = {
  IMAGE: ['image/jpeg', 'image/png', 'image/webp'],
  DOCUMENT: [
    'application/pdf',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  ],
  AUDIO: [
    'audio/wav',
    'audio/x-wav',
    'audio/wave',
    'audio/mpeg',
    'audio/mp3',
    'audio/mp4',
    'audio/x-m4a',
    'audio/aac',
    'audio/ogg',
    'audio/webm',
  ],
};

// WHY: Keep event names in one place to avoid typos.
const CHAT_SOCKET_EVENTS = {
  MESSAGE_NEW: 'message:new',
  MESSAGE_READ: 'message:read',
  CONVERSATION_JOIN: 'conversation:join',
  CONVERSATION_LEAVE: 'conversation:leave',
  CALL_INCOMING: 'call:incoming',
  CALL_UPDATED: 'call:updated',
  CALL_SIGNAL: 'call:signal',
  CALL_ENDED: 'call:ended',
  ERROR: 'chat:error',
};

module.exports = {
  CHAT_CONVERSATION_TYPES,
  CHAT_MESSAGE_TYPES,
  CHAT_CALL_MEDIA_MODES,
  CHAT_CALL_STATES,
  CHAT_ATTACHMENT_TYPES,
  CHAT_ATTACHMENT_MIME_TYPES,
  CHAT_LIMITS,
  CHAT_SOCKET_EVENTS,
};

File: apps/frontend/lib/app/features/home/presentation/chat_constants.dart
/// lib/app/features/home/presentation/chat_constants.dart
/// ------------------------------------------------------
/// WHAT:
/// - Shared constants for chat UI + socket events.
///
/// WHY:
/// - Prevents magic strings across chat widgets and services.
/// - Keeps event names aligned with backend expectations.
///
/// HOW:
/// - Exposes event names, message types, and UI copy tokens.
library;

// WHY: Socket event names must match backend chat constants.
const String chatEventMessageNew = "message:new";
const String chatEventMessageRead = "message:read";
const String chatEventConversationJoin = "conversation:join";
const String chatEventConversationLeave = "conversation:leave";
const String chatEventCallIncoming = "call:incoming";
const String chatEventCallUpdated = "call:updated";
const String chatEventCallSignal = "call:signal";
const String chatEventCallEnded = "call:ended";
const String chatEventError = "chat:error";

// WHY: Message types map to rendering logic.
const String chatMessageTypeText = "text";
const String chatMessageTypeAttachment = "attachment";
const String chatMessageTypeSystem = "system";

// WHY: Conversation types map to UI labels.
const String chatConversationTypeDirect = "direct";
const String chatConversationTypeGroup = "group";

// WHY: Call states are reused across overlays and socket updates.
const String chatCallStateRinging = "ringing";
const String chatCallStateActive = "active";
const String chatCallStateDeclined = "declined";
const String chatCallStateEnded = "ended";
const String chatCallStateMissed = "missed";
const String chatCallStateCancelled = "cancelled";

// WHY: Audio is enabled now; video remains a future-compatible option.
const String chatCallMediaModeAudio = "audio";
const String chatCallMediaModeVideo = "video";
